server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -184,11 +192,12 @@ func (e *EventChannel) MarkInvocationDisconnected(ctx context.Context, iid strin
 	}
 
 	ti := tableInvocationFromProto(invocation, iid)
-	if cacheStats := hit_tracker.CollectCacheStats(e.ctx, e.env, iid); cacheStats != nil {
-		fillInvocationFromCacheStats(cacheStats, ti)
-	}
 	recordInvocationMetrics(ti)
-	return e.env.GetInvocationDB().InsertOrUpdateInvocation(ctx, ti)
+	if err := e.env.GetInvocationDB().InsertOrUpdateInvocation(ctx, ti); err != nil {
+		return err
+	}
+	go e.writeCacheStatsToDBWhenReady(context.Background(), iid)
+	return nil
```

Comment:
Maybe a better approach here would be to throw these invocation ids down a channel, and have something on the other end handling them once they are ready. This would allow you to also register a graceful shutdown function that closed and drained this queue, so you know these get processed when a server restarts, rather than possibly dropped.

Reply:
Not sure I understand how these would get dropped, or why the risk of being dropped is lower with the queue + worker approach? Once the goroutine is started it can't be canceled just because the server is shutting down right?

Reply:
You said it! It can and will be cancelled as soon aS the main function ends, which is exactly why you want to put these in a channel. Then you can have a single goroutine that works on these, and you can call RegisterShutdownFunction and drain that queue, which will keep main from exiting until you've finished.


One way to think about what you have now is a go routine leak. At high qps you're going to be spinning up hundreds of these and taking up n*qps resources. If you put them in a queue instead you'll have lighter resource usage and better behavior during shutdown / rollout.

Reply:
Done.